src/pointer-analysis/dereference.cpp

```diff
@@ -338,138 +347,70 @@ void dereferencet::dereference_addrof_expr(
   dereference_expr(expr, guard, mode);
 }
 
-void dereferencet::dereference_deref(expr2tc &expr, guardt &guard, modet mode)
-{
-  if(is_dereference2t(expr))
-  {
-    dereference2t &deref = to_dereference2t(expr);
-    // first make sure there are no dereferences in there
-    dereference_expr(deref.value, guard, dereferencet::READ);
-
-    expr2tc tmp_obj = deref.value;
-    expr2tc result = dereference(tmp_obj, deref.type, guard, mode, expr2tc());
-    expr = result;
-  }
-  else
-  {
-    // This is an index applied to a pointer, which is essentially a dereference
-    // with an offset.
-    assert(is_index2t(expr) && is_pointer_type(to_index2t(expr).source_value));
-    std::list<expr2tc> scalar_step_list;
-    assert(
-      (is_scalar_type(expr) || is_code_type(expr)) &&
-      "Can't dereference to a nonscalar type");
-    index2t &idx = to_index2t(expr);
-
-    // first make sure there are no dereferences in there
-    dereference_expr(idx.index, guard, dereferencet::READ);
-    dereference_expr(idx.source_value, guard, mode);
-
-    add2tc tmp(idx.source_value->type, idx.source_value, idx.index);
-    // Result discarded.
-    expr = dereference(tmp, tmp->type, guard, mode, expr2tc());
-  }
-}
-
 expr2tc dereferencet::dereference_expr_nonscalar(
   expr2tc &expr,
   guardt &guard,
   modet mode,
-  std::list<expr2tc> &scalar_step_list)
+  const expr2tc &base)
 {
   if(is_dereference2t(expr))
   {
+    /* The first expression we're called with is index2t, member2t or non-scalar
+     * if2t. Thus, expr differs from base. */
+    assert(expr != base);
+
+    // Check that either the base type that these steps are applied to matches
+    // the type of the object we're wrapping in these steps. It's a type error
+    // if there isn't a match.
+    type2tc base_of_steps_type = ns.follow(expr->type);
+
+    if(dereference_type_compare(expr, base_of_steps_type))
```

Comment:
Can't you remove the else and negate the condition? `if(!dereference_type_compare(expr, base_of_steps_type))`?

Reply:
OK, done